x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java

```diff
@@ -54,26 +57,56 @@
     public static final String ILM_HISTORY_INDEX_PREFIX = "ilm-history-" + INDEX_TEMPLATE_VERSION + "-";
     public static final String ILM_HISTORY_ALIAS = "ilm-history-" + INDEX_TEMPLATE_VERSION;
 
-    private final Client client;
-    private final ClusterService clusterService;
     private final boolean ilmHistoryEnabled;
     private final BulkProcessor processor;
+    private final ThreadPool threadPool;
 
-    public ILMHistoryStore(Settings nodeSettings, Client client, ClusterService clusterService) {
-        this.client = client;
-        this.clusterService = clusterService;
-        ilmHistoryEnabled = LIFECYCLE_HISTORY_INDEX_ENABLED_SETTING.get(nodeSettings);
+    public ILMHistoryStore(Settings nodeSettings, Client client, ClusterService clusterService, ThreadPool threadPool) {
+        this.ilmHistoryEnabled = LIFECYCLE_HISTORY_INDEX_ENABLED_SETTING.get(nodeSettings);
+        this.threadPool = threadPool;
 
         this.processor = BulkProcessor.builder(
             new OriginSettingClient(client, INDEX_LIFECYCLE_ORIGIN)::bulk,
             new BulkProcessor.Listener() {
                 @Override
-                public void beforeBulk(long executionId, BulkRequest request) { }
+                public void beforeBulk(long executionId, BulkRequest request) {
+                    // Prior to actually performing the bulk, we should ensure the index exists, and
+                    // if we were unable to create it or it was in a bad state, we should not
+                    // attempt to index documents.
+                    try {
+                        final CompletableFuture<Boolean> indexCreated = new CompletableFuture<>();
+                        ensureHistoryIndex(client, clusterService.state(), ActionListener.wrap(indexCreated::complete,
+                            ex -> {
+                                logger.warn("failed to create ILM history store index prior to issuing bulk request", ex);
+                                indexCreated.completeExceptionally(ex);
+                            }));
+                        try {
+                            indexCreated.get();
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException("interrupted waiting for ILM history index to be created prior to indexing", e);
+                        } catch (ExecutionException e) {
+                            throw new RuntimeException("cannot index ILM history items as index could not be created", e);
+                        }
```

Comment:
As we're rethrowing and re-catching these exceptions below would it be more readable to call `indexCreated.join()` instead of `get()` and let the exceptions bubble to the below `catch` block? (as we're rethrowing them anyway and I think the original InterruptedException/ExecutionException.cause would be expressive enough to point to what happened)